### PR TITLE
vm: clean up the external interface

### PR DESCRIPTION
**Pull request by zerbina**
## Summary

Invoking the VM was previously done through an `execute` procedure that was mainly meant for use by the compiler, as it had on-demand procedure code-gen (via `vmjit`) and VM-data deserialization built in. The final goal is to completely decouple the VM from the rest of compiler, and breaking this point of coupling is another step towards this goal.

To encapsulate the state of a single thread of execution, the `VmThread` object is introduced. `initVmThread` creates a new instance; `execute` runs a thread until it yields; and `dispose` cleans up the resources used by the thread and destroys it.

`ExecResult` is renamed to `YieldReason` and VM errors are now also reported through the `YieldReason` object. The previous `vmjit` integration as well the VM-data -> ``PNode`` translation are moved into a procedure in the `compilerbridge` module that is implemented on top of the `VmThread` primitive.

In addition, the `opcQuit` instruction no longer implements the quit logic itself, but yields execution with a dedicated reason, leaving how a `quit` call is to be handled up the VM's callsite. This is done as an experiment to see if a similar, but more generalized mechanism could in the future be used to replace the current VM callbacks.

As a direct consequence, the `vm` no longer imports `vmjit` (which imported `vmgen`).

## Details
- adjust the `vmrunner` to work with the VM API
- export `createStackTrace` and use a `VmThread` for providing the PC and stack frame position
- document issues with `opcGorge`, `opcEcho`, and the way ``quit`` works at compile-time

### CODE REVIEW

File: compiler/vm/compilerbridge.nim [hunk 1]
```diff
@@ -49,14 +50,27 @@ import
...
+  ExecErrorReport* = object
+    stackTrace*: VMReport ## The report storing the stack-trace
+    report*: VMReport     ## The report detailing the error
+
+  ExecutionResult* = Result[PNode, ExecErrorReport]
+
 # to prevent endless recursion in macro instantiation
 const evalMacroLimit = 1000
 
+# prevent a default `$` implmenentation from being generated
```
Comment (saem) [inline compiler/vm/compilerbridge.nim]:
```suggestion
# prevent a default `$` implementation from being generated
```

File: compiler/vm/compilerbridge.nim [hunk 2]
```diff
@@ -138,6 +152,73 @@ proc unpackResult(res: sink ExecutionResult; config: ConfigRef, node: PNode): PN
...
+      # execution is finished
+      result.initSuccess cb(c, c.sframes[0].slots[r.reg.get])
+      break
+    of yrkError:
+      result.initFailure buildError(c, thread, r.error)
+      break
+    of yrkQuit:
+      case c.mode
+      of emRepl, emStaticExpr, emStaticStmt:
+        # XXX: should code run at compile time really be able to force-quit
```
Comment (saem) [inline compiler/vm/compilerbridge.nim:193]: That seems... really really weird. I'm trying to think of a use case that's not awful and I can't. All I can think of is how it doesn't help composition.

File: compiler/vm/vm.nim [hunk 3]
```diff
@@ -9,6 +9,13 @@
 
 ## This file implements the new evaluation engine for Nim code.
 ## An instruction is 1-3 int32s in memory, it is a register based VM.
+## # TODO: update the above -- it doesn' reflect reality anymore
```
Comment (saem) [inline compiler/vm/vm.nim]:
```suggestion
## # TODO: update the above -- it doesn't reflect reality anymore
```

File: compiler/vm/vm.nim [hunk 4]
```diff
@@ -9,6 +9,13 @@
...
+##
+## To use the VM once has to first set up an execution environment, which is
+## represented by ``TCtx``. In order to execute something, a ``VmThread``
```
Comment (saem) [inline compiler/vm/vm.nim:15]: I kinda want to call it a 'process' instead of 'thread' as we might want some sort of green thread/continuation thing.

**Review by saem — APPROVED** [review]
Minor suggestions mostly.

I'm not entirely convinced about 'process' but I'm a big fan of how things are decoupling either way. This weekend is going to be really exciting to come back to the code. 🎉

File: compiler/vm/vm.nim (hunk 4 above)
Comment (zerbina) [reply compiler/vm/vm.nim:15]: I originally wanted to name it `VmInstance` (and before that, `ExecState`), but ultimately decided on `VmThread`.

In my opinion, 'process' sounds a bit heavy-weight for what the object actually represents (the execution state for a single thread), but I'm not strongly against renaming it to `VmProcess`.

Comment (saem) [reply compiler/vm/vm.nim:15]: Let's leave it as is. We can always rename it later and have the compiler help us do it.

**Comment by saem** [comment]
Bors r+

**Pull request merged**